synapseclient/core/multithread_download/download_threads.py

```diff
@@ -0,0 +1,393 @@
+    ...
+
+    Attributes
+    ----------
+    file_name: str
+        name of the file for the presigned url
+    url: str
+        the actual presigned url
+    expiration_utc: datetime.datetime
+        datetime in UTC at which the url will expire
+    """
+    file_name: str
+    url: str
+    expiration_utc: datetime.datetime
+
+class PresignedUrlProvider(object):
+    """
+    Provides an un-exipired pre-signed url to download a file
+    """
+    request: DownloadRequest
+    _cached_info: PresignedUrlInfo
+
+
+    # offset parameter used to buffer url expiration checks, time in seconds
+    _TIME_BUFFER: datetime.timedelta = datetime.timedelta(seconds=5)
+
+    def __init__(self, client, request: DownloadRequest):
+        self.client = client
+        self.request: DownloadRequest = request
+        self._cached_info: PresignedUrlInfo = self._get_pre_signed_info()
+        self._lock = _threading.Lock()
+
+    def get_info(self) -> PresignedUrlInfo:
+        with self._lock:
+            if datetime.datetime.utcnow() + PresignedUrlProvider._TIME_BUFFER >= self._cached_info.expiration_utc:
+                self._cached_info = self._get_pre_signed_info()
+
+            return self._cached_info
+
+    def _get_pre_signed_info(self) -> PresignedUrlInfo:
+        """
+        Returns the file_name and pre-signed url for download as specified in request
+
+        :return: PresignedUrlInfo
+        """
+        # noinspection PyProtectedMember
+        response = self.client._getFileHandleDownload(self.request.file_handle_id, self.request.object_id)
+        file_name = response["fileHandle"]["fileName"]
+        pre_signed_url = response["preSignedURL"]
+        return PresignedUrlInfo(file_name, pre_signed_url, _pre_signed_url_expiration_time(pre_signed_url))
+
+
+class DataChunkDownloadThread(_threading.Thread):
+    """
+    The producer threads that make the GET request and obtain the data for a download chunk
+    """
+    def __init__(self, presigned_url_provider: PresignedUrlProvider, range_queue:CloseableQueue, data_queue:CloseableQueue):
+        super().__init__()
+        self.daemon = True
+        self.presigned_url_provider = presigned_url_provider
+        self.range_queue = range_queue
+        self.data_queue = data_queue
+        self.session = _get_new_session()
+
+    def run(self):
+        for byte_range in self.range_queue:
+            start, end = byte_range
+
+            response = self._get_response_with_retry(start, end)
+
+            try:
+                for bytes in response.iter_content(MAX_CHUNK_WRITE_SIZE):
+                    self.data_queue.put((start, bytes))
+                    start += len(bytes)
+            except QueueClosedException:
+                # the data_queue was closed so stop retrieving data chunks
+                response.close()
+                break
+
+    def _get_response_with_retry(self, start:int, end:int) -> Response:
+        range_header = {'Range': f'bytes={start}-{end}'}
```

Comment:
Just checking to see if we are supporting python 3.6+?  f-interpolation doesn't work in python 3.5.

Reply:
we are supporting 3.6+ since the latest version is 3.8.

Reply:
Will need to make sure this is updated: https://github.com/Sage-Bionetworks/synapsePythonClient/blob/develop/setup.py#L35